Pull request baseball game

bohub12 opened the pull request:
Computer.java를 싱글턴패턴을 써서 메모리 낭비 방지
일급 컬렉션을 구현하긴했지만, 써야되는 이유와는 동떨어진 감이 있는것같다 리팩토링하며 생각해보자
클래스가 하나의 패키지 안에있는데 리팩토링하며 클래스기능에 맞게 패키지별로 분리해야할것같다
로직이 잘 돌아가는지 확인하는 테스트 코드를 짜볼 수 없을까

--------
좀 더 가독성 있게끔 Naming 하기
static과 final의 이점 등 어디에 유용하게 쓰이는지 알아보기
예외 처리에 대한 이해 부족. 좀 더 알아볼것

bohub12 committed:
feat(Computer.java) : getBallCount, getStrikeCount event

bohub12 committed:
feat(Computer.java) : printCount, printPlaying, printEndGame event related output functions

bohub12 committed:
feat(Computer.java) : restartOrNot event

bohub12 committed:
feat(Computer.java) : createRandomNum event

This method creates a number with different three digits.

bohub12 committed:
feat(Computer.java) : filterReceivedNum event

This method returns proper received number.
But if inappropriate number is entered, it returns false.

bohub12 committed:
feat(User.java) : play event

This method prints the process of playing once
for one random number.

bohub12 committed:
feat(Application.java) : program

bohub12 committed:
test(ComputerTest.java)

It tests creating random number, counting ball count and strike count,
inserting appropriate number by user

bohub12 committed:
feat : Computer 기능 세분화

bohub12 committed:
feat : 재시작과 관련된 제어기능 구현

bohub12 committed:
feat : 게임 돌아가는 구조 구현

bohub12 committed:
feat : 사용자 입력이 유효한지 확인하는 기능

bohub12 committed:
feat : 3개의 임의의 수로 난수 만드는 일급 컬렉션 구현

bohub12 committed:
feat : 결과 출력 기능 구현

bohub12 committed:
feat : 한 난수에 대해 플레이하는 구조 구현

bohub12 committed:
test : 기능 세분화 및 카운트 관련 테스트 코드

bohub12 committed:
test : 3자리 각기 다른 난수 생성에 관한 테스트 코드

bohub12 committed:
test : 정해진 범위 내에 숫자 생성에 관한 테스트 코드

bohub12 committed:
test : 범위 내외의 숫자 들어올때의 상황에 관한 테스트코드

bohub12 committed:
test : 재시작 관련 범위 내외의 숫자 들어올 때의 테스트

bohub12 committed:
style : 테스트 코드위해 코드 수정

wansoo05 commented on src/main/java/baseball/InputHandler.java:
```diff
@@ -0,0 +1,21 @@
+        }
+        return true;
+    }
+
```
사용자가 입력한 숫자에 중복검사가 없습니다. 정답이 123인경우 111을 입력할 시에 2볼 1스트라이크라는 잘못된 흰트가 나올 수 있기 때문에 중복검사구현을 하면 좋을것 같습니다. 그리고, 1~9 만사용이가능한데 0을 사용해도 오류가 없습니다. 이것또한 추가 구현하면 좋을것 같습니다.

wansoo05 replied:
추가로 개인적인 의견인데 전체적으로 잘못된 입력값을 맏을때 IllegalArumentException을 발생시키고 다시 입력받게 하는 방법도 좋은것같습니다

bohub12 replied:
위에 리뷰는 고려하지못햇네요
그나저나 exception발생시키고 어떻게 다시 입력을 받을 수 있죠?? exception을 발생시키면 종료되지않나요??

wansoo05 replied:
저는 try catch문으로 미리 예외를 대비해서 예외가 발생하여 받으면 다시 입력받도록 했습니다

wansoo05 commented on src/main/java/baseball/Computer.java:
```diff
@@ -0,0 +1,39 @@
+package baseball;
+import utils.RandomUtils;
+
```
RandomUtils가 사용되고 있지 않으므로 import를 remove하는게 좋을것 같습니다.

bohub12 commented on src/main/java/baseball/GameLogic.java:
```diff
@@ -0,0 +1,16 @@
+
+public class GameLogic {
+    public static void run() {
+        Controller.setGAMEOVER(false);
```
setter보단 Controller에 메서드를 추가해서 GAMEOVER를 설정하는게 더 좋을 것 같다. (리팩토링에 반영)

bohub12 committed:
fix : 볼 계산시 StringIndexOutofBoundsException발생 버그 고침

bohub12 committed:
refactor : setter없애고 메서드 통해 상수 상태바꾸기

bohub12 committed:
style : 패키지 구조 변경과 변경사항에 따른 코드 변경

ingyuseong commented on src/main/java/baseball/Application.java:
```diff
@@ -5,6 +5,11 @@
 public class Application {
     public static void main(String[] args) {
         final Scanner scanner = new Scanner(System.in);
```
`scanner` 변수를 왜 선언하셨나요?

ingyuseong commented on src/main/java/baseball/Computer.java:
```diff
@@ -0,0 +1,38 @@
+package baseball;
+
+public class Computer {
```
음 마찬가지로 저는 Computer라는 클래스 명이 되게 직관적이지 못하다라고 생각합니다. 좋지 못한 naming이라고 생각합니다.

+ 음 강의에서 싱글턴을 이렇게 구현했었던 것 같은데, static 메소드로 싱글턴 안만들고 이런 식으로 구현했을 경우에 이점이 있나요?

bohub12 replied:
음 전 static 메서드로 싱글턴을 만들엇는데?? 어떤 걸 말하는 거죠??

ingyuseong replied:
그럼 Computer.java 소스코드 확인해보시면 static 안쓰셨어요!

ingyuseong commented on src/main/java/baseball/Computer.java:
```diff
@@ -0,0 +1,38 @@
+        return strikeCount;
+    }
+
+    public String goString(int num) {
```
ㅋㅋ 함수명 naming 조금 더 신경써야하지 않나 싶습니다. `goString`..? 이름만 봤을 때도 아 대충 무슨 일하겠구나라는 감이 잡히는 이름이 좋은 이름인 것 같습니다.

ingyuseong commented on src/main/java/baseball/Controller.java:
```diff
@@ -0,0 +1,53 @@
+
+import java.util.Scanner;
+
+public class Controller {
```
Naming!

ingyuseong commented on src/main/java/baseball/GameLogic.java:
```diff
@@ -0,0 +1,16 @@
+            User user = new User();
+            user.play(random);
+
+            Controller.restartOrNot(Controller.getRestartNumber());
```
에러 발생시키고, 에러 핸들링을 안해줬네요.

ingyuseong commented on src/main/java/baseball/Controller.java:
```diff
@@ -0,0 +1,53 @@
+        } else if(n == END) {
+            Controller.stop();
+        } else {
+            throw new IllegalArgumentException();
```
이거 맞나요..? raise가 들어가는 것 아닌가요?

bohub12 replied:
raise가 들어간다는게 무슨말이져?

ingyuseong replied:
raise 예약어에 대해 구글링해보시면 될 것 같습니다 :) 저도 아직 예외처리 쪽이 미흡해서 정확하진 않아요.

ingyuseong commented on src/main/java/baseball/NumberCreator.java:
```diff
@@ -0,0 +1,35 @@
+    public void createValidSet() {
+        while (!isValidSize(randomNumbers)) {
+            randomNumbers.add(RandomUtils.nextInt(1, 9));
+        }
```
저라면 do while쓸 것 같습니다. 그게 의도가 더 잘 드러나는 것 같습니다.

set 활용한 것 아주 좋습니다.

ingyuseong commented on src/main/java/baseball/NumberCreator.java:
```diff
@@ -0,0 +1,35 @@
+    public final Set<Integer> randomNumbers = new HashSet<>();
+
+    public boolean isValidSize(Set<Integer> randomNumbers) {
+        if (randomNumbers.size() == 3) {
+            return true;
+        }
+        return false;
```
#### Reference
* [객체 지향 생활체조 규칙 2. else 예약어 금지](https://developerfarm.wordpress.com/2012/01/27/object_calisthenics_3/)
* [Null Object Pattern](https://johngrib.github.io/wiki/null-object-pattern/)

ingyuseong commented on src/main/java/baseball/User.java:
```diff
@@ -0,0 +1,31 @@
+                THREESTRIKE = true;
+            }
+        } while (!THREESTRIKE);
+    }
```
전체적으로 코딩 컨벤션 어긴 부분이 꽤 많이 보입니다. 저도 확인했는데도 실수로 몇번 어긴경우가 있어서, 말씀안드리고 있었는데, 이건 심합니다 ㅠ

ingyuseong commented on src/main/java/baseball/User.java:
```diff
@@ -0,0 +1,31 @@
+        } while (!THREESTRIKE);
+    }
+
+    public int insert() {
```
Naming!

ingyuseong commented on src/main/java/baseball/verification/InputHandler.java:
```diff
@@ -0,0 +1,32 @@
+package baseball.verification;
+
+public class InputHandler {
```
싱글톤 의도하신거면,, 인스턴스 생성 제한해두시는게 좋을 것 같습니다.

bohub12 replied:
인스턴스 생성 막는게 좋겟네 감사해요

ingyuseong commented on src/main/java/baseball/verification/OutputHandler.java:
```diff
@@ -0,0 +1,50 @@
+
+    private static String NOTHING = "낫싱 ";
+    private static String BALL = "볼 ";
+    private static String STRIKE = "스트라이크 ";
```
final!

ingyuseong commented on src/main/java/baseball/verification/OutputHandler.java:
```diff
@@ -0,0 +1,50 @@
+        System.out.println(resultCount(ballCount, strikeCount).toString());
+    }
+
+    public static void printStart() {
+        System.out.print(MESSAGE);
+    }
+
+    public static void printEnd() {
+        System.out.println(ENDMESSAGE);
+        System.out.println(REMESSAGE);
+    }
```
음... 저는 이거 맞는 건지 잘 모르겠습니다. 취향 차이일 수도 있고, 이거에 관한 best practice가 있는지는 잘모르겠지만, 한번 찾아보는 걸 추천드립니다.

ingyuseong commented on src/main/java/baseball/verification/OutputHandler.java:
```diff
@@ -0,0 +1,50 @@
+        System.out.println(REMESSAGE);
+    }
+
+    public static StringBuilder resultCount (int ballCount, int strikeCount) {
```
Naming.

ingyuseong commented on src/main/java/baseball/verification/OutputHandler.java:
```diff
@@ -0,0 +1,50 @@
+
+import baseball.Computer;
+
+public class OutputHandler {
```
이름을 왜 OutputHandler라고 하셨나요?

bohub12 replied:
이 클래스에선 출력과 관련된 메서드들만 선언해놓아서 출력과 관련된 클래스라는 의미로 outputhandler라 햇죠

bohub12 committed:
refactor : 세부사항 밑에서 다룬다.
최대한 코드 컨벤션 따르기(else 예약어 금지 등)
예외처리하고나서도 프로그램 반복해서 이어갈 수 있게끔 리팩토링
Naming 가독성 신경쓰며 메서드 클래스 이름 바꾸기
원시값과 원시열 등 static final로 포장하기
유연한 설계 되게끔 클래스의 기능 분리 및 메서드 기능 분리
